modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAtomicLongDataStructureTest.java

```diff
@@ -0,0 +1,180 @@
+    /** Nodes. */
+    protected static final int NODES_CNT = 4;
+
+    /** Keys count. */
+    protected static final int KEYS_CNT = 100;
+
+    /** Corrupted keys count. */
+    protected static final int BROKEN_KEYS_CNT = 10;
+
+    /** Data structure name. */
+    protected static final String DS_NAME = "DefaultDS";
+
+    /** Fix mode. */
+    @Parameterized.Parameter(0)
+    public boolean fixMode;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String name) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(name);
+
+        cfg.setDataStorageConfiguration(new DataStorageConfiguration()
+            .setDefaultDataRegionConfiguration(new DataRegionConfiguration()
+                .setPersistenceEnabled(true)
+                .setMaxSize(300L * 1024 * 1024))
+        );
+
+        cfg.setConsistentId(name);
+        cfg.setAutoActivationEnabled(false);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        stopAllGrids();
+
+        cleanPersistenceDir();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+
+        cleanPersistenceDir();
+    }
+
+    /**
+     *
+     */
+    @Parameterized.Parameters(name = "repair = {0}")
+    public static List<Object[]> parameters() {
+        ArrayList<Object[]> params = new ArrayList<>();
+        params.add(new Object[] {false});
+        params.add(new Object[] {true});
+        return params;
+    }
+
+    /**
+     * Tests that reconciliation works with atomic long.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    @WithSystemProperty(key = IGNITE_SENSITIVE_DATA_LOGGING, value = "plain")
+    public void testReconciliationOfAtomicLong() throws Exception {
+        String cacheName = "ignite-sys-atomic-cache@default-ds-group";
+
+        IgniteEx ig = startGrids(NODES_CNT);
+
+        IgniteEx client = startClientGrid(NODES_CNT);
+
+        client.cluster().active(true);
+
+        AtomicConfiguration cfg = new AtomicConfiguration().setBackups(NODES_CNT - 1);
+
+        for (int i = 0; i < KEYS_CNT; i++)
+            client.atomicLong(DS_NAME + i, cfg, 1, true);
+
+        log.info(">>>> Initial data loading finished");
+
+        int firstBrokenKey = KEYS_CNT / 2;
+
+        GridCacheContext[] nodeCacheCtxs = new GridCacheContext[NODES_CNT];
+
+        for (int i = 0; i < NODES_CNT; i++)
+            nodeCacheCtxs[i] = grid(i).cachex(cacheName).context();
+
+        for (int i = firstBrokenKey; i < firstBrokenKey + BROKEN_KEYS_CNT; i++) {
+            Class<?> aCls = Class.forName(GridCacheInternalKeyImpl.class.getCanonicalName());
+            for (Constructor<?> constructor : aCls.getDeclaredConstructors()) {
+                constructor.setAccessible(true);
+                if (constructor.getParameterCount() == 2) {
```

Comment:
Better to find the constructor before the loop instead of looking for it for each key. Also, please add a comment with the signature you expect.

Also, why is this needed? I see both GridCacheInternalKeyImpl constructors are public